gensim/models/poincare.py

```diff
@@ -0,0 +1,831 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Author: Jayant Jain <jayantjain1992@gmail.com>
+# Copyright (C) 2017 Radim Rehurek <me@radimrehurek.com>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+
+"""
+Python implementation of Poincare Embeddings [1], an embedding that is better at capturing latent hierarchical
+information better than traditional Euclidean embeddings. The method is described in more detail in [1].
+
+The main use-case is to automatically learn hierarchical representations of nodes from a tree-like structure,
+such as a Directed Acyclic Graph, using the transitive closure of the relations.
+
+This module allows training a Poincare Embedding from a training file containing relations from
+a transitive closure.
+
+[1] Maximilian Nickel, Douwe Kiela - "Poincaré Embeddings for Learning Hierarchical Representations"
+    https://arxiv.org/pdf/1705.08039.pdf
+
+Examples
+--------
+Initialize and train a model from a list::
+
+>>> from gensim.models.poincare import PoincareModel
+>>> relations = [('kangaroo', 'marsupial'), ('kangaroo', 'mammal'), ('gib', 'cat')]
+>>> model = PoincareModel(relations, negative=2)
+>>> model.train(epochs=50)
+
+Initialize and train a model from a file containing one relation per line::
+
+>>> from gensim.models.poincare import PoincareModel, PoincareRelations
+>>> file_path = 'gensim/test/test_data/poincare_hypernyms.tsv'
+>>> model = PoincareModel(PoincareRelations(file_path), negative=2)
+>>> model.train(epochs=50)
+
+"""
+
+
+import csv
+import logging
+import sys
+import time
+
+import numpy as np
+from collections import defaultdict, Counter
+from numpy import random as np_random
+from smart_open import smart_open
+
+from gensim import utils
+from gensim.models.keyedvectors import KeyedVectors, Vocab
+
+
+logger = logging.getLogger(__name__)
+
+
+class PoincareModel(utils.SaveLoad):
+    """Class for training, using and evaluating Poincare Embeddings.
+
+    The model can be stored/loaded via its `save()` and `load()` methods, or stored/loaded in the word2vec
+    format via `wv.save_word2vec_format()` and `KeyedVectors.load_word2vec_format()`.
+
+    Note that training cannot be resumed from a model loaded via `load_word2vec_format`, if you wish to train further,
+    use `save()` and `load()` methods instead.
+
+    """
+    def __init__(
+        self, train_data, size=50, alpha=0.1, negative=10, workers=1,
+        epsilon=1e-5, burn_in=10, burn_in_alpha=0.01, init_range=(-0.001, 0.001), seed=0):
+        """Initialize and train a Poincare embedding model from an iterable of transitive closure relations.
```

Comment:
Is the transitive closure a requirement? If not, let's just say "iterable of relations".

Reply:
Fixed.